parlai/utils/testing.py

```diff
@@ -307,14 +326,41 @@ def eval_model(opt, skip_valid=False, skip_test=False):
         popt['dict_file'] = popt['model_file'] + '.dict'
 
     with capture_output() as output:
-        popt['datatype'] = 'valid'
+        popt['datatype'] = 'valid' if valid_datatype is None else valid_datatype
         valid = None if skip_valid else ems.eval_model(popt)
         popt['datatype'] = 'test'
         test = None if skip_test else ems.eval_model(popt)
 
     return (output.getvalue(), valid, test)
 
 
+def eval_model_timeout(
+    opt, skip_valid=False, skip_test=False, valid_datatype=None, timeout_seconds=None
+):
+    """
+    Run through evaluation loop, with timeout.
+
+    :param opt:
+        Any non-default options you wish to set.
+    :param bool skip_valid:
+        If true skips the valid evaluation, and the second return value will be None.
+    :param bool skip_test:
+        If true skips the test evaluation, and the third return value will be None.
+    :param str valid_datatype:
+        If custom datatype required for valid, e.g. train:evalmode, specify here
+    :param int timeout:
+        Seconds to wait before timeout. Fails if None
+
+    :return: (stdout, valid_results, test_results)
+    :rtype: (str, dict, dict)
+    """
+    if timeout_seconds is None or timeout_seconds <= 0:
+        raise TypeError('Timeout must be positive')
+
+    with timeout(timeout_seconds):
```

Comment:
how would you feel about this being folded into the test directly, rather than this small helper method?

Reply:
yeah that's probably better